lightning/src/chain/channelmonitor.rs

```diff
@@ -2920,6 +2928,27 @@ pub trait Persist<ChannelSigner: Sign> {
 	/// [`ChannelMonitorUpdate::write`] for writing out an update, and
 	/// [`ChannelMonitorUpdateErr`] for requirements when returning errors.
 	fn update_persisted_channel(&self, id: OutPoint, update: &ChannelMonitorUpdate, data: &ChannelMonitor<ChannelSigner>) -> Result<(), ChannelMonitorUpdateErr>;
+
+	/// Update one channel's data synchronously without a [`ChannelMonitorUpdate`].
+	///
+	/// This is called during block/transaction connection, and is a good time to synchronously
+	/// remove all pending [`ChannelMonitorUpdate`]s which may have been persisted separately as an
+	/// intent log.
+	///
+	/// Note that returning an error here irrevocably disables some processing in [`ChainMonitor`],
+	/// preventing continued normal operation. Errors here are largely only useful to continue
+	/// operation long enough to shut down.
+	///
+	/// Failures here do not imply the channel will be force-closed, however any future calls to
+	/// [`update_persisted_channel`] after an error is returned here MUST either persist the full,
+	/// updated [`ChannelMonitor`] provided to [`update_persisted_channel`] or return
+	/// [`ChannelMonitorUpdateErr::PermanentFailure`], force-closing the channel. In other words,
+	/// any future calls to [`update_persisted_channel`] after an error here MUST NOT persist the
+	/// [`ChannelMonitorUpdate`] alone.
```

Comment:
Could we simplify this by making `update_persisted_channel` take an `Option<&ChannelMonitorUpdate>` and forgoing adding `sync_persisted_channel` method to the trait? Then when the user sees `None` they must write the full `ChannelMonitor`.

Reply:
The return value distinction is the harder part to capture - users don't really have the option to return a `PermanentFailure` here as the call site (the `ChainMonitor`) doesn't have the ability to force-close a channel.

Reply:
Yea, that's a good point, I guess I hadn't thought about making the update an option and skipping it if we've marked a channel as "previous sync didn't persist". The "don't call `update_persisted_channel`"  change, though feels wrong - technically if the user complies with the API docs we can continue operating fine, after a restart, but if we don't call `update_persisted_channel` at all we have to force-close the channel immediately.

Reply:
What if after a block comes in, we set a bool in `ChannelMonitor` indicating there’s been a new block? Then we can change `update_persisted_channel` to pass in whether a full sync is needed to avoid a big chain resync. 

Can also check the new bool in `timer_tick` and call `update_persisted_channel` then, if we also make the `ChannelMonitorUpdate` an `Option` maybe

Reply:
We still need to sync the monitor before we can hand any chain-generated events back to the ChannelManager, so I don't think we can avoid a sync at all.